modules/core/shared/src/test/scala/retry/PackageObjectSpec.scala

```diff
@@ -216,6 +216,192 @@ class PackageObjectSpec extends AnyFlatSpec {
     assert(gaveUp)
   }
 
+  behavior of "retryingOnFailuresAndSomeErrors"
+
+  it should "retry until the action succeeds" in new TestContext {
+    val policy = RetryPolicies.constantDelay[StringOr](1.second)
+
+    val finalResult = retryingOnFailuresAndSomeErrors[String](
+      policy,
+      _.exists(_.toInt > 3),
+      (_: String) == "one more time",
+      onError,
+      onError
+    ) {
+      attempts = attempts + 1
+      if (attempts < 3)
+        Left("one more time")
+      else
+        Right("yay")
+    }
+
+    assert(finalResult == Right("yay"))
+    assert(attempts == 3)
+    assert(errors.toList == List("one more time", "one more time"))
+    assert(!gaveUp)
+  }
+
+  it should "retry only if the error is worth retrying" in new TestContext {
+    val policy = RetryPolicies.constantDelay[StringOr](1.second)
+
+    val finalResult = retryingOnFailuresAndSomeErrors[String](
+      policy,
+      _.exists(_.toInt > 3),
+      (_: String) == "one more time",
+      onError,
+      onError
+    ) {
+      attempts = attempts + 1
+      if (attempts < 3)
+        Left("one more time")
+      else
+        Left("nope")
+    }
+
+    assert(finalResult == Left("nope"))
+    assert(attempts == 3)
+    assert(errors.toList == List("one more time", "one more time"))
+    assert(!gaveUp) // false because onError is only called when the error is worth retrying
+  }
+
+  it should "retry until the policy chooses to give up" in new TestContext {
+    val policy = RetryPolicies.limitRetries[StringOr](2)
+
+    val finalResult = retryingOnFailuresAndSomeErrors[String](
+      policy,
+      _.exists(_.toInt > 3),
+      (_: String) == "one more time",
+      onError,
+      onError
+    ) {
+      attempts = attempts + 1
+      Left("one more time")
+    }
+
+    assert(finalResult == Left("one more time"))
+    assert(attempts == 3)
+    assert(
+      errors.toList == List("one more time", "one more time", "one more time")
+    )
+    assert(gaveUp)
+  }
+
+  it should "retry in a stack-safe way" in new TestContext {
+    val policy = RetryPolicies.limitRetries[StringOr](10000)
+
+    val finalResult = retryingOnFailuresAndSomeErrors[String](
+      policy,
+      _.exists(_.toInt > 3),
+      (_: String) == "one more time",
+      onError,
+      onError
+    ) {
+      attempts = attempts + 1
+      Left("one more time")
+    }
+
+    assert(finalResult == Left("one more time"))
+    assert(attempts == 10001)
+    assert(gaveUp)
+  }
```

Comment:
All these tests are testing the handling of errors, i.e. `Left`. Please could you add a test to check the failure case, i.e. the action returns a `Right` but the `wasSuccessful` predicate fails.

Reply:
Will do!